parasol_cpu/src/proc/ops/store.rs

```diff
@@ -1,117 +1,72 @@
     unwrap_registers,
 };
 
-use super::{check_offset, read_write_mask};
-
 impl FheProcessor {
     #[allow(clippy::too_many_arguments)]
     /// Execute a store instruction.
     pub fn store(
         &mut self,
         retirement_info: RetirementInfo<DispatchIsaOp>,
-        scoreboard_entry: ScoreboardEntryRef<DispatchIsaOp>,
+        memory: &Memory,
         src: RobEntryRef<Register>,
-        dst: RobEntryRef<PtrRegister>,
-        width: u32,
+        dst: RobEntryRef<Register>,
+        num_bytes: u32,
         instruction_id: usize,
-        pc: usize,
+        pc: u32,
     ) {
-        let store_impl = |scoreboard_entry: &ScoreboardEntryRef<DispatchIsaOp>| -> Result<()> {
+        let store_impl = || -> Result<()> {
             let mut dst = dst.entry_force_mut();
             let dst = &mut **dst;
 
             unwrap_registers!((src));
 
-            let mask = read_write_mask(width);
-            let num_bytes = width.next_multiple_of(8) as usize / 8;
-
-            match (src, dst) {
-                (
-                    Register::Plaintext {
-                        val,
-                        width: reg_width,
-                    },
-                    PtrRegister::Plaintext(ptr),
-                ) => {
-                    check_offset(width, ptr.offset, ptr.base.len(), instruction_id, pc)?;
-
-                    if *reg_width != width {
-                        return Err(Error::WidthMismatch {
-                            inst_id: instruction_id,
-                            pc,
-                        });
-                    }
-
-                    // Plaintext loads and stores exec immediately, so we don't need to update our deps.
-
-                    for i in 0..(num_bytes - 1) {
-                        let shift_amt = 8 * i;
-                        *ptr.base[i].borrow_mut() = (val >> shift_amt) as u8;
-                    }
-
-                    let shift_amt = 8 * (num_bytes - 1);
-                    *ptr.base[num_bytes - 1].borrow_mut() = mask & (val >> shift_amt) as u8;
+            match dst {
+                Register::Plaintext { val: ptr, width: _ } => {
+                    let base_addr = *ptr as u32;
 
-                    FheProcessor::retire(&retirement_info, Ok(()));
-                }
-                (
-                    Register::Ciphertext(val),
-                    PtrRegister::Ciphertext(CiphertextPtr::PlainOffset(ptr)),
-                ) => {
-                    check_offset(width, ptr.offset, ptr.base.len(), instruction_id, pc)?;
-
-                    if val.len() != width as usize {
-                        return Err(Error::WidthMismatch {
-                            inst_id: instruction_id,
-                            pc,
-                        });
+                    if base_addr % num_bytes != 0
+                        || num_bytes > 16
+                        || !num_bytes.is_power_of_two()
+                        || num_bytes == 0
```

Comment:
Since these conditions are reused between load and store it might be good to abstract them into one function in case we need to change the conditions.